TypeError: type object argument after ** must be a mapping, not Sphinx

Subject: TypeError: type object argument after ** must be a mapping, not Sphinx

### Problem
- Cant use sphinx because the sphinxcontrib.googleanalytics package throws an error.

#### Procedure to reproduce the problem
```
make html
sphinx-build -b html -d build/doctrees   source build/html
```
#### Error logs / results

##### Python 3.6.3
```
$ sphinx-build -b html -d build/doctrees   source build/html
Running Sphinx v1.7.2
WARNING: extension 'sphinxcontrib.googleanalytics' returned an unsupported object from its setup() function; it should return None or a metadata dictionary

Exception occurred:
  File "/usr/local/lib/python3.6/dist-packages/sphinx/registry.py", line 344, in load_extension
    app.extensions[extname] = Extension(extname, mod, **metadata)
TypeError: type object argument after ** must be a mapping, not Sphinx
The full traceback has been saved in /tmp/sphinx-err-c4n8npty.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
Makefile:55: recept voor doel 'html' is mislukt
make: *** [html] Fout 2
```

```
# Sphinx version: 1.7.2
# Python version: 3.6.3 (CPython)
# Docutils version: 0.14 
# Jinja2 version: 2.10
# Last messages:

# Loaded extensions:
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/sphinx/cmdline.py", line 303, in main
    args.warningiserror, args.tags, args.verbosity, args.jobs)
  File "/usr/local/lib/python3.6/dist-packages/sphinx/application.py", line 191, in __init__
    self.setup_extension(extension)
  File "/usr/local/lib/python3.6/dist-packages/sphinx/application.py", line 411, in setup_extension
    self.registry.load_extension(self, extname)
  File "/usr/local/lib/python3.6/dist-packages/sphinx/registry.py", line 344, in load_extension
    app.extensions[extname] = Extension(extname, mod, **metadata)
TypeError: type object argument after ** must be a mapping, not Sphinx
```

##### Python 2.7.12
```
$ python2.7 /usr/local/bin/sphinx-build -b latex -d build/doctrees   source build/latex
Running Sphinx v1.7.2
WARNING: extension 'sphinxcontrib.googleanalytics' returned an unsupported object from its setup() function; it should return None or a metadata dictionary

Exception occurred:
  File "/usr/local/lib/python2.7/dist-packages/sphinx/registry.py", line 344, in load_extension
    app.extensions[extname] = Extension(extname, mod, **metadata)
TypeError: type object argument after ** must be a mapping, not Sphinx
The full traceback has been saved in /tmp/sphinx-err-AqY4hJ.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
```

```
# Sphinx version: 1.7.2
# Python version: 2.7.12 (CPython)
# Docutils version: 0.14 
# Jinja2 version: 2.10
# Last messages:

# Loaded extensions:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/sphinx/cmdline.py", line 303, in main
    args.warningiserror, args.tags, args.verbosity, args.jobs)
  File "/usr/local/lib/python2.7/dist-packages/sphinx/application.py", line 191, in __init__
    self.setup_extension(extension)
  File "/usr/local/lib/python2.7/dist-packages/sphinx/application.py", line 411, in setup_extension
    self.registry.load_extension(self, extname)
  File "/usr/local/lib/python2.7/dist-packages/sphinx/registry.py", line 344, in load_extension
    app.extensions[extname] = Extension(extname, mod, **metadata)
TypeError: type object argument after ** must be a mapping, not Sphinx
```

### Reproducible project / your project
- https://github.com/nickeubank/data-analysis-in-python

### Environment info
- OS: Kernel: x86_64 Linux 4.10.0-22-generic
- Python version: Python 3.6.3 & 2.7.12
- Sphinx version: sphinx-build 1.7.2

--- a/CHANGES
+++ b/CHANGES
@@ -32,6 +32,7 @@ Bugs fixed
 * toctree directive tries to glob for URL having query_string
 * #4871: html search: Upper characters problem in German
 * #4717: latex: Compilation for German docs failed with LuaLaTeX and XeLaTeX
+* #4878: Crashed with extension which returns invalid metadata
 
 Testing
 --------

--- a/sphinx/registry.py
+++ b/sphinx/registry.py
@@ -340,6 +340,7 @@ class SphinxComponentRegistry(object):
             logger.warning(__('extension %r returned an unsupported object from '
                               'its setup() function; it should return None or a '
                               'metadata dictionary'), extname)
+            metadata = {}
 
         app.extensions[extname] = Extension(extname, mod, **metadata)
         app._setting_up_extension.pop()

